In [1]:
import pandas as pd
import numpy as np

Desafío 1: Generación de funciones
- Genere funciones para calcular la media y varianza de un vector. Debe cumplir con
los siguientes requisitos:
  - Ambas funciones deben ingresar un argumento x correspondiente al vector.
  - Las funciones deben contener docstrings con la documentación asociada a
la variable y retornar un valor utilizando return.
  - La función de la varianza se debe llamar a la función de la media.
- Utilice las funciones para reportar la información sobre goles_favor, goles_contra
y puntos.


In [2]:
def calcular_media(vector):
    """Función para calcular la media de un vector (una lista-array)

    :param vector: Es el vector-lista-array ingresado a la función
    :type list:

    :return: Retorna la media del vector
    :rtype: float64
    """
    suma = 0
    for i in vector:
        suma += i

    media = suma / len(vector)

    return media

def calcular_varianza(vector):
    """Función para calcular la varianza de un vector (una lista-array)

    :param vector: Es el vector-lista-array ingresado a la función
    :type list:

    :return: Retorna la varianza del vector
    :rtype: float64
    """
    media = calcular_media(vector)

    suma = 0
    for i in vector:
        suma += ((i - media) ** 2)
    
    varianza = suma / len(vector)

    return varianza

In [3]:
df = pd.read_csv('worldcup2014.csv')

In [4]:
columnas_a_analizar = ['goles_favor', 'goles_contra', 'puntos']
for columna in columnas_a_analizar:
    media = calcular_media(df[columna])
    varianza = calcular_varianza(df[columna])

    print(f'Analizando la columna "{columna}"\n'
          f'Media: {media}\n'
          f'Varianza: {varianza}\n')


Analizando la columna "goles_favor"
Media: 4.25
Varianza: 5.1875

Analizando la columna "goles_contra"
Media: 4.25
Varianza: 4.8125

Analizando la columna "puntos"
Media: 4.21875
Varianza: 8.0458984375



Desafío 2:

Utilizando el método groupby de la clase DataFrame, en conjunto con la función .agg de
pandas, calcule la media, la varianza y desviación estándar de la cantidad de goles a favor,
en contra y de la cantidad de puntos por continente.

In [5]:
df.groupby('continent')[columnas_a_analizar].agg([calcular_media, calcular_varianza,
                                                  pd.DataFrame.std])


goles_favor                               goles_contra  \
             calcular_media calcular_varianza       std calcular_media   
continent                                                                
africa             3.600000          2.640000  1.816590       5.600000   
asia               2.250000          0.687500  0.957427       6.250000   
europe             4.769231          6.331361  2.618964       4.000000   
northamerica       3.250000          1.687500  1.500000       3.500000   
southamerica       5.666667          3.888889  2.160247       2.833333   

                                                 puntos                    \
             calcular_varianza       std calcular_media calcular_varianza   
continent                                                                   
africa                3.840000  2.190890       2.400000          2.640000   
asia                  3.187500  2.061553       0.750000          0.187500   
europe                3.538462  1.957890       4.692308          6.366864   
northamerica          8.250000  3.316625       4.500000          8.250000   
southamerica          0.472222  0.752773       6.833333          3.138889   

                        
                   std  
continent               
africa        1.816590  
asia          0.500000  
europe        2.626297  
northamerica  3.316625  
southamerica  1.940790

- ¿En qué continente se observa una mayor cantidad de goles a favor? En sudamérica se observa
una mayor cantidad de goles a favor en promedio (media de 5.66, con una desviacion estandar de 2.16)

- ¿En qué continente se observa una mayor cantidad de goles en contra? En asia se observa una mayor
cantidad de goles en contra en promedio (media de 6.25, con una desviación estándar de 2.06)

- ¿En qué continente se observa una mayor cantidad de puntos en promedio? En sudamerica se observa
una mayor cantidad de puntos en promedio (media de 6.83, con una desviación estándar de 1.94)

Desafío 3: Simulaciones
- Genere una función generate_pet que devuelva de forma aleatoria un string
'perro' o 'gato' un número n de veces. Ejecútela un par de veces.
Tip: Puede utilizar la función np.random.choice para retornar elementos al azar.

In [14]:
def generate_pet(numero):
    """Función para generar una lista con los strings aleatorios 'perro' o 'gato'

    :param numero: Es la cantidad de elementos que va a tener la lista retornada
    :type: int

    :return: Retorna una lista con una cantidad numero de strings aleatorios 'perro' o 'gato'
    :rtype: list
    """
    opciones = ['perro', 'gato']

    lista_contenedora = []
    for i in range(numero):
        eleccion = np.random.choice(opciones)
        lista_contenedora.append(eleccion)
    
    return lista_contenedora

print(generate_pet(1))
print(generate_pet(5))
print(generate_pet(10))
print(generate_pet(30))

['perro']
['gato', 'perro', 'perro', 'gato', 'perro']
['perro', 'gato', 'gato', 'perro', 'gato', 'gato', 'gato', 'gato', 'perro', 'perro']
['perro', 'gato', 'perro', 'gato', 'perro', 'gato', 'perro', 'perro', 'perro', 'gato', 'gato', 'perro', 'perro', 'gato', 'gato', 'gato', 'gato', 'gato', 'gato', 'perro', 'perro', 'perro', 'perro', 'gato', 'gato', 'gato', 'gato', 'perro', 'gato', 'perro']


- Aplique la función generate_pet para generar 20 muestras.

In [7]:
muestra_20 = generate_pet(20)

- ¿Cuál es la probabilidad de elegir un perro al azar? ¿Y un gato?

In [8]:
prob_perro = muestra_20.count('perro') / len(muestra_20)
prob_gato = muestra_20.count('gato') / len(muestra_20)

print(f'La probabilidad de elegir a un perro en una muestra de 20 es de: {prob_perro}\n'
      f'La probabilidad de elegir a un gato en una muestra de 20 es de: {prob_gato}')

La probabilidad de elegir a un perro en una muestra de 20 es de: 0.15
La probabilidad de elegir a un gato en una muestra de 20 es de: 0.85


- Agregue np.random.seed(2) al inicio del chunk. ¿Qué diferencia hay cuando se
ejecuta la función varias veces luego de fijar la semilla?

In [17]:
def generate_pet(numero):
    """Función para generar una lista con los strings aleatorios 'perro' o 'gato'. Tiene un 
    random seed de 2

    :param numero: Es la cantidad de elementos que va a tener la lista retornada
    :type: int

    :return: Retorna una lista con una cantidad numero de strings aleatorios 'perro' o 'gato'
    :rtype: list
    """
    opciones = ['perro', 'gato']

    np.random.seed(2)
    lista_contenedora = []
    for i in range(numero):
        eleccion = np.random.choice(opciones)
        lista_contenedora.append(eleccion)
    
    return lista_contenedora

In [18]:
muestra_20 = generate_pet(20)
prob_perro = muestra_20.count('perro') / len(muestra_20)
prob_gato = muestra_20.count('gato') / len(muestra_20)

print(f'La probabilidad de elegir a un perro en una muestra de 20 es de: {prob_perro}\n'
      f'La probabilidad de elegir a un gato en una muestra de 20 es de: {prob_gato}')

La probabilidad de elegir a un perro en una muestra de 20 es de: 0.4
La probabilidad de elegir a un gato en una muestra de 20 es de: 0.6


Al fijar la semilla se genera siempre el mismo estado random (siempre se generan la misma cantidad
de gatos/perros, y en el mismo orden siempre). Por lo tanto, las probabilidades de sacar a un perro
o un gato son constantes.

Desafío 4: Función simuladora
- Genere una función llamada simulate_pets_prob que tome como argumento un
número finito de simulaciones a generar.
- La función debe simular dos situaciones young_pet y old_pet, y contar la
ocurrencia de los siguientes casos:
  - De las dos mascotas simuladas (young y old), contar las ocasiones donde
por lo menos una de las mascotas sea un perro.
  - De las dos mascotas simuladas, contar las ocasiones donde old_pet sea un
perro.
  - De las dos mascotas simuladas, contar las ocasiones donde los dos sean
perros.
- El método debe tener una semilla pseudoaleatoria de 1.
- El output de la función debe entregarse en términos de Probabilidad.


In [15]:
def simulate_pets_prob(numero_simulaciones):
    """Función que simula dos situaciones que generan una lista de animales aleatoria. Utiliza
    la función generate_pet para generar la lista de animales aleatorias. Además, calcula
    3 probabilidades distintas.

    :param numero_simulaciones: Es la cantidad de animales que van a ser generados por la función
    generate_pet
    :type: int

    :return: Nada
    :rtype: None
    """

    np.random.seed(1)
    young_pet = generate_pet(numero_simulaciones)
    old_pet = generate_pet(numero_simulaciones)

    df = pd.DataFrame(data={'young_pet': young_pet, 'old_pet': old_pet})

    df_tiene_perro = df.query('young_pet == "perro" or old_pet == "perro"')
    cantidad_tiene_perro = df_tiene_perro.shape[0]
    prob_tiene_perro = cantidad_tiene_perro / numero_simulaciones

    df_old_pet_perro = df.query('old_pet == "perro"')
    cantidad_old_pet_perro = df_old_pet_perro.shape[0]
    prob_old_pet_perro = cantidad_old_pet_perro / numero_simulaciones

    df_ambos_perro = df.query('young_pet == "perro" and old_pet == "perro"')
    cantidad_ambos_perro = df_ambos_perro.shape[0]
    prob_ambos_perro = cantidad_ambos_perro / numero_simulaciones

    print(f'La probabilidad de que young_pet u old_pet tengan un perro es del: {prob_tiene_perro}')
    print(f'La probabilidad de que old_pet tenga un perro es del: {prob_old_pet_perro}')
    print(f'La probabilidad de que ambos tengan un perro es del: {prob_ambos_perro}')


In [16]:
simulate_pets_prob(100)

La probabilidad de que young_pet u old_pet tengan un perro es del: 0.75
La probabilidad de que old_pet tenga un perro es del: 0.54
La probabilidad de que ambos tengan un perro es del: 0.24


- De los tres escenarios, ¿Cuál es el menos probable? ¿Cuál es el más probable? ¿Por
qué?

- El escenario menos probable es que young_pet y old_pet tengan "perro". El escenario más probable
es que young_pet u old_pet tengan "perro". 

1. El primer escenario mencionado es mucho menos probable
que ocurra debido a que es una intersección de eventos (que young_pet **Y** old_pet tengan el mismo 
resultado), por lo que la cantidad de observaciones de tal evento es menor. 
2. Por otro lado, el
segundo evento mencionado es mayor, ya que corresponde a una unión de eventos (que young_pet 
**O** old_pet tengan a perro), por lo que la cantidad de observaciones de tal evento es mayor.